doc/prov_loadex.md

```diff
@@ -0,0 +1,83 @@
+Providers run-time configuration
+================================
+
+Currently any provider run-time activation requires presence of the
+initialization parameters in the OpenSSL configuration file. Otherwise the
+provider will be activated with some "default" settings, that may or may not
+work for a particular application. For real-world systems it may require
+providing a specially designed OpenSSL config and passing it somehow (e.g. via
+environment) that has its obvious drawbacks.
+
+We need a possibility to initialize providers on per-application level
+according to per-application parameters. It's necessary for example for PKCS#11
+provider (where different applications may use different devices with different
+drivers) and will be useful for some other providers. In case of Red Hat it is
+also usable for FIPS provider.
+
+Real-world cases
+----------------
+
+Many applications use PKCS#11 API with a specific drivers. OpenSSL PKCS#11
+provider (https://github.com/latchset/pkcs11-provider) also provide a set of
+tweaks usable in particular situations. So there are at least several scenarios
+I have in mind:
+
+1. Configure a provider in the congig file, activate on demand
+2. Partially configure a provider in the config file, provide some extra
+   parameters later
+3. Load/activate a provider run-time and provide parameters later
+4. Load/activate a provider run-time with parameters
+5. Load a provider with a specific config snippet (usability)
+
+Proposed API
+------------
+
+All the names suggested below can be changed. I'm pretty bad in naming.
+
+#Configure a provider in the congig file, activate on demand
+
+We have an existing API `OSSL_PROVIDER_load`
+
+#Partially configure a provider in the config file, provide some extra
+parameters later
+
+We need an extra function
+
+```C
+int OSSL_PROVIDER_set_params(OSSL_PROVIDER *prov, OSSL_PARAM params[]);
+```
+
+and probably a syntax sugar
+
+```C
+int OSSL_PROVIDER_set_params_by_name(OSSL_LIB_CTX *libctx, const char *name,
```

Comment:
My intention here is the scenario: we have loaded the provider and specified most of the parameters (via config) but not all, and we want to specify the remaining ones.

Reply:
As @t8m commented, it will be too late.
Once the provider is initialized any change in params will be ignored.
We would have to add a provider API and change providers to provide these function to update parameters after initialization.

Reply:
OK, I got your point.